src/plonk_recursion.rs

```diff
@@ -101,15 +109,18 @@ pub fn recursive_verification_circuit<C: HaloCurve, InnerC: HaloCurve<BaseField=
     // Can call curve_assert_valid.
 
     // Compute random challenges.
-    let mut challenger = RecursiveChallenger::new();
+    let mut challenger = RecursiveChallenger::<C>::new();
     challenger.observe_affine_points(&proof.c_wires);
     let (beta, gamma) = challenger.get_2_challenges(&mut builder);
     challenger.observe_affine_point(proof.c_plonk_z);
     let alpha = challenger.get_challenge(&mut builder);
     challenger.observe_affine_points(&proof.c_plonk_t);
     let zeta = challenger.get_challenge(&mut builder);
     challenger.observe_elements(&proof.all_opening_targets());
-    let (v, u, x) = challenger.get_3_challenges(&mut builder);
+    let (v, u) = challenger.get_2_challenges(&mut builder);
+
+    // Challenge for scaling u used in the IPA verification.
+    let u_scaling = challenger.get_challenge(&mut builder);
```

Comment:
Since we don't have prover messages between (v, u) and u_scaling, I think it would be best to keep using `get_3_challenges` just for performance (i.e. our current sponge width is 4, so we can get 3 outputs with a single permutation).

Reply:
Or we could keep it like this and modify Challenger to store an output buffer (and discard it if the transcript changes). Maybe that would be better since it looks like `schnorr_challenge` could also be combined with `l_challenge` below, but that might make the code awkward. I could work on that later.

Reply:
I'll change it to `get_3_challenges` for now.